zcash_client_sqlite/src/wallet.rs

```diff
@@ -166,35 +194,65 @@ pub(crate) fn memo_repr(memo: Option<&MemoBytes>) -> Option<&[u8]> {
     })
 }
 
-pub(crate) fn get_max_account_id(
+// Returns the highest used account index for a given seed.
+pub(crate) fn max_zip32_account_index(
     conn: &rusqlite::Connection,
-) -> Result<Option<AccountId>, SqliteClientError> {
-    // This returns the most recently generated address.
-    conn.query_row_and_then("SELECT MAX(account) FROM accounts", [], |row| {
-        let account_id: Option<u32> = row.get(0)?;
-        account_id
-            .map(AccountId::try_from)
-            .transpose()
-            .map_err(|_| SqliteClientError::AccountIdOutOfRange)
-    })
+    seed_id: &HdSeedFingerprint,
+) -> Result<Option<zip32::AccountId>, SqliteClientError> {
+    conn.query_row_and_then(
+        "SELECT MAX(hd_account_index) FROM accounts WHERE hd_seed_fingerprint = :hd_seed",
+        [seed_id.as_bytes()],
+        |row| {
+            let account_id: Option<u32> = row.get(0)?;
+            account_id
+                .map(zip32::AccountId::try_from)
+                .transpose()
+                .map_err(|_| SqliteClientError::AccountIdOutOfRange)
+        },
+    )
+}
+
+fn get_sql_values_for_account_parameters<'a, P: consensus::Parameters>(
```

Comment:
This kind of direct 1:1 mapping really makes me think that the `Account` enum should just be an internal detail of `zcash_client_sqlite`, at least until we have a better documented idea of what general purpose should have in the public API (rather than what I am now seeing it as having reviewed the `zcash_client_sqlite` part of the PR, which is "Rust representation of our internal SQL columns, leaking implementation details").

Reply:
And just in case I'm coming across as very negative here: I'd be very happy to be convinced of the utility of the `Account` enum in the public API! I'm also very tired of the amount of churn our APIs have been going through for a while now, and I know of downstream users who are as well. Add in the time pressure to get Orchard implemented, and I really want to minimize new public API changes.

What I want is for new centrally-important APIs to have a cohesive vision. If that is what `Account` is being proposed as, I'd love for a new issue to be opened in which the full extent of its usage is conveyed.

Reply:
We already have `get_account_birthday`, which retrieves just one field for a given account. How about we replace that with `get_account` and add birthday to the struct? Then the struct can represent all the scalar values we know about an account.

Reply:
I removed the API for this PR.

Reply:
I think adding `birthday` would likely make sense (per my original comments about birthday being a necessary detail for an account). This then means `Account` is indeed more of a "general metadata about an account" type, and also that if `Account` is proposed as _the way_ for accessing per-account information rather than per-attribute methods, then we should refactor more of the `WalletRead` trait methods around it.